src/datatypes/dataset.ts

```diff
@@ -1067,46 +1077,31 @@ export class Dataset implements Instance<DatasetValue, DatasetJS> {
     });
   }
 
-  public fullJoin(other: Dataset, compare: (v1: any, v2: any) => number): Dataset {
+  public fullJoin(other: Dataset): Dataset {
     if (!other || !other.data.length) return this;
     const { data, keys, attributes } = this;
     if (!data.length) return other;
 
-    let thisKey = keys[0]; // ToDo: temp fix
-    if (!thisKey) throw new Error('join lhs must have a key (be a product of a split)');
-    if (thisKey !== other.keys[0]) throw new Error('this and other keys must match');
+    if (!this.sameKeys(other)) {
+      throw new Error('this and other keys must match');
+    }
 
-    const otherData = other.data;
-    const dataLength = data.length;
-    const otherDataLength = otherData.length;
-    let newData: Datum[] = [];
-    let i = 0;
-    let j = 0;
-    while (i < dataLength || j < otherDataLength) {
-      if (i < dataLength && j < otherDataLength) {
-        const nextDatum = data[i];
-        const nextOtherDatum = otherData[j];
-        const cmp = compare(nextDatum[thisKey], nextOtherDatum[thisKey]);
-        if (cmp < 0) {
-          newData.push(nextDatum);
-          i++;
-        } else if (cmp > 0) {
-          newData.push(nextOtherDatum);
-          j++;
+    const myDatumLookup = this.getKeyLookup();
+    const otherDatumLookup = other.getKeyLookup();
+
+    const newData = deduplicateSort(Object.keys(myDatumLookup).concat(Object.keys(otherDatumLookup))).map(keyValue => {
```

Comment:
Why call this `keyValue` instead of just `key`?

Reply:
you are right, it should just be key